x-pack/legacy/plugins/siem/public/pages/detection_engine/components/signals/helpers.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { get, isEmpty } from 'lodash/fp';
+import { esKuery } from '../../../../../../../../../src/plugins/data/common';
+import { esFilters } from '../../../../../../../../../src/plugins/data/public';
+import {
+  DataProvider,
+  DataProvidersAnd,
+} from '../../../../components/timeline/data_providers/data_provider';
+import { Ecs } from '../../../../graphql/types';
+
+interface FindValueToChangeInQuery {
+  field: string;
+  valueToChange: string;
+}
+
+const templateFields = [
+  'host.name',
+  'host.hostname',
+  'host.domain',
+  'host.id',
+  'host.ip',
+  'client.ip',
+  'destination.ip',
+  'server.ip',
+  'source.ip',
+  'network.community_id',
+  'user.name',
+  'process.name',
+];
+
+export const findValueToChangeInQuery = (
+  keuryNode: esKuery.KueryNode,
+  valueToChange: FindValueToChangeInQuery[] = []
+): FindValueToChangeInQuery[] => {
+  let localValueToChange = valueToChange;
+  if (keuryNode.function === 'is' && templateFields.includes(keuryNode.arguments[0].value)) {
+    localValueToChange = [
+      ...localValueToChange,
+      {
+        field: keuryNode.arguments[0].value,
+        valueToChange: keuryNode.arguments[1].value,
+      },
+    ];
+  }
+  return keuryNode.arguments.reduce(
+    (addValueToChange: FindValueToChangeInQuery[], ast: esKuery.KueryNode) => {
+      if (ast.function === 'is' && templateFields.includes(ast.arguments[0].value)) {
+        return [
+          ...addValueToChange,
+          {
+            field: ast.arguments[0].value,
+            valueToChange: ast.arguments[1].value,
+          },
+        ];
+      }
+      if (ast.arguments) {
+        return findValueToChangeInQuery(ast, addValueToChange);
+      }
+      return addValueToChange;
+    },
+    localValueToChange
+  );
+};
+
+export const replaceTemplateFieldFromQuery = (query: string, ecsData: Ecs) => {
+  if (query !== '') {
+    const valueToChange = findValueToChangeInQuery(esKuery.fromKueryExpression(query));
+    return valueToChange.reduce((newQuery, vtc) => {
+      const newValue = get(vtc.field, ecsData);
+      if (newValue != null) {
+        return newQuery.replace(vtc.valueToChange, newValue);
+      }
+      return newQuery;
+    }, query);
+  }
+  return '';
+};
+
+export const replaceTemplateFieldFromMatchFilters = (filters: esFilters.Filter[], ecsData: Ecs) => {
+  if (!isEmpty(filters)) {
```

Comment:
If it's already an empty array this doesn't do any performance gain. The first thing map, filter, etc... do is check if the array is empty and return an empty array back for you so you don't have to add extra blocks of if checks and indentation.